subsys/net/lib/sockets/sockets.c

```diff
@@ -1416,11 +1477,44 @@ static ssize_t zsock_recv_stream_timed(struct net_context *ctx, uint8_t *buf, si
 			}
 		}
 
-		res = zsock_recv_stream_immediate(ctx, &buf, &max_len, flags);
-		recv_len += res;
-		if (res == 0) {
-			if (recv_len == 0 && K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
-				return -EAGAIN;
+		if (msg != NULL) {
+			res = zsock_recv_stream_immediate(ctx, &buf, &available_len, flags);
+			recv_len += res;
+
+			if (res == 0) {
+				if (recv_len == 0 && K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
+					return -EAGAIN;
+				}
+			}
+
+			msg->msg_iov[iovec].iov_len += res;
+			buf = (uint8_t *)(msg->msg_iov[iovec].iov_base) + res;
+			max_len -= res;
+
+			if (available_len == 0) {
+				/* All data to this iovec was written */
+				iovec++;
+
+				if (iovec == max_iovlen) {
+					break;
+				}
+
+				msg->msg_iovlen = iovec;
+				buf = msg->msg_iov[iovec].iov_base;
+				available_len = msg->msg_iov[iovec].iov_len;
+				msg->msg_iov[iovec].iov_len = 0;
+
+				continue;
```

Comment:
Should it also be checked if there is any remaining data on the rx queue before `continue`? Because right now, if the data available on the stream matched exactly `iov_len`, and there is more `msg_iov` available, we're getting back to `zsock_wait_data()`, potentially blocking while the function could return what was received so far.